### PR TITLE
[PLAT-920][Hotfix] Implement user deletion protocol 

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

We need a GDPR compliant delete method, this gives us one.

## Changes

- add method gdpr_delete to user model
- adds management command wrapper for method
- a whole bunch of tests.

## QA Notes

Use the management command `python manage.py gdpr_delete_user guid1` to delete different users and see that all their identifying info is striped and that there are no references to them on the site. If you attempt to delete a user with preprints or registrations or a user who is an admin on a shared project, the delete should fail.

## Documentation

This feature is for internal use only.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/PLAT-920

### CODE REVIEW

File: osf/models/user.py [hunk 1]
```diff
@@ -1508,6 +1508,56 @@ def get_node_comment_timestamps(self, target_id):
...
+        user_nodes = self.nodes.all()
+        #  Validates the user isn't trying to delete things they deliberately made public.
+        if user_nodes.filter(type='osf.registration').exists():
+            raise ValidationError('You cannot delete this user because they have one or more registrations.')
+
+        if PreprintService.objects.filter(node___contributors=self, ever_public=True).exists():
+            raise ValidationError('You cannot delete this user because they have one or more preprints.')
+
+        # Validates that the user isn't trying to delete things nodes they are the only admin on.
+        personal_nodes = [node.id for node in user_nodes if node.contributor_set.count() == 1]
```
Comment (Johnetordoff) [inline osf/models/user.py]: This is the best I could up with, it's hard to count the length of a queryset collect via through table. So I was unable to to this via annotation, not sure if it would be possible or how best to do it, but this works and doesn't really need to be optimized.

File: osf_tests/test_user.py [hunk 2]
```diff
@@ -1968,3 +1970,79 @@ def test_validate_schools_bad_year(self):
...
+        second_admin_contrib = UserFactory()
+        project = ProjectFactory(creator=self.user)
+        project.add_contributor(second_admin_contrib)
+        project.set_permissions(user=second_admin_contrib, permissions=['read', 'write', 'admin'])
+        project.save()
+
+        self.user.gdpr_delete()
+
+        # The deleted user is still associated with the node, though their name still appears as 'Deleted User'
+        assert self.user.nodes.all().count() == 1
```
Comment (Johnetordoff) [inline osf_tests/test_user.py]: This case where a user is deleted, but still technically a contributor could be problematic, but we just have to trust the remaining admins to remove deleted contributors when appropriate.

File: osf_tests/test_user.py [hunk 3]
```diff
@@ -1968,3 +1970,79 @@ def test_validate_schools_bad_year(self):
             }]
             with pytest.raises(ValidationError):
                 self.user.save()
+
+
+class TestUserGdprDelete(OsfTestCase):
```
Comment (sloria) [inline osf_tests/test_user.py]: Drive-by: New tests should use plain classes and pytest fixtures instead of OsfTestCase and unittest style.

File: osf/models/user.py (hunk 1 above)
Comment (sloria) [reply osf/models/user.py]: Does this work?

```python
AbstractNode.objects.annotate(contrib_count=Count('_contributors')).filter(contrib_count__lte=1).filter(contributor__user=me)
```

You'll probably need to add more filtering (filtering out bookmarks, quickfiles etc.) but you get the idea.

File: osf/models/user.py [hunk 4]
```diff
@@ -1508,6 +1508,56 @@ def get_node_comment_timestamps(self, target_id):
...
+        This function does not remove the user object reference from our database, but it does disable the account and
+        remove identifying in a manner compliant with GDPR guidelines.
+        :return:
+        """
+        from osf.models import PreprintService, AbstractNode
+
+        user_nodes = self.nodes.all()
+        #  Validates the user isn't trying to delete things they deliberately made public.
+        if user_nodes.filter(type='osf.registration').exists():
+            raise ValidationError('You cannot delete this user because they have one or more registrations.')
```
Comment (sloria) [inline osf/models/user.py]: `ValidationError` is used when model field validation fails, so I don't think it's quite right here. A custom error might make sense, e.g. `UserStateError` (for consistency with `NodeStateError`).

File: osf/models/user.py [hunk 5]
```diff
@@ -1508,6 +1508,56 @@ def get_node_comment_timestamps(self, target_id):
...
+
+        if PreprintService.objects.filter(node___contributors=self, ever_public=True).exists():
+            raise ValidationError('You cannot delete this user because they have one or more preprints.')
```
Comment (caseyrollins) [inline osf/models/user.py]: Drive-by: this script should handle _deactivating_ users that have preprints and/or registrations.

File: osf/models/user.py [hunk 6]
```diff
@@ -1508,6 +1508,56 @@ def get_node_comment_timestamps(self, target_id):
...
+                    'You cannot delete node {} because it would be a node with contributors, but with no admin.'.format(
+                        node._id))
+
+        # delete all personal nodes (one contributor), bookmarks, quickfiles etc.
+        personal_nodes.delete()
+
+        # This removes identifying info
+        self.fullname = 'Deleted User'
+        self.username = 'Deleted User'
+        self.given_name = 'Deleted User'
```
Comment (caseyrollins) [inline osf/models/user.py]: Drive-by: I would suggest setting `fullname` to "Deleted User" and setting `given`, `family`, and `middle` to an empty string. There may be place where we display names as given + family which would lead to "Deleted User Deleted User" in this case.

Comment (caseyrollins) [reply osf/models/user.py]: And let's clear `suffix` while we're at it.

Comment (caseyrollins) [reply osf/models/user.py]: I haven't looked into these two fields to see how they're used, but I would think we should clear `mailchimp_mailing_lists` and `osf_mailing_lists` as well so that a mailing list doesn't end up with a null email.

File: osf/models/user.py (hunk 1 above)
Comment (Johnetordoff) [reply osf/models/user.py]: That will actually work fine verbatim, we do want to remove Quickfiles ect, that's part of the requirements.

File: osf/models/user.py (hunk 5 above)
Comment (Johnetordoff) [reply osf/models/user.py]: @caseyrollins I think admins can already deactivate users manually if they get this message. I'm not sure it helps to deactivate them automatically. If the user really wants to delete themselves and they have to change ownership of a shared project first they won't be able to do it if they're deactivated.

File: osf/models/user.py [hunk 7]
```diff
@@ -1508,6 +1509,58 @@ def get_node_comment_timestamps(self, target_id):
...
+        self.given_name = ''
+        self.family_name = ''
+        self.middle_names = ''
+        self.mailchimp_mailing_lists = {}
+        self.osf_mailing_lists = {}
+        self.suffix = ''
+        self.jobs = []
+        self.schools = []
+        self.social = []
+        self.external_accounts.clear()
```
Comment (sloria) [inline osf/models/user.py]: Drive-by: this just detaches the external_accounts from the user but doesn't actually redact the information, correct? IIRC, we want to scrub out all the fields on associated external accoutns.
